File: java/Builder/com/fastercoding/user/UserBuilder.java
package com.fastercoding.user;

import com.fastercoding.user.User;

public class UserBuilder {

    String firstName;
    String lastName;
    int age;
    String street;

    private UserBuilder() {

    }

    public static UserBuilder user() {
        return new UserBuilder();
    }

    public UserBuilder withFirstName(String name) {
        this.firstName = name;
        return this;
    }

    public UserBuilder withLastName(String name) {
        this.lastName = name;
        return this;
    }

    public UserBuilder withAge(int age) {
        this.age = age;
        return this;
    }

    public UserBuilder liveInStreet(String street) {
        this.street = street;
        return this;
    }

    public User build() {
        User user = new User(this.firstName,
                             this.lastName,
                             this.age,
                             this.street);
        
        return user;
    }

}

File: typescript/Builder/src/user/user-builder.ts
import {User} from "./user";

export class UserBuilder {

    private _firstName: string;
    private _lastName: string;
    private _age: number;
    private _street: string;

    private constructor() {
        this._firstName = "";
        this._lastName = "";
        this._age = -1;
        this._street = "";
    }

    public static user(): UserBuilder {
        return new UserBuilder();
    }

    public withFirstName(name: string): UserBuilder {
        this._firstName = name;
        return this;
    }

    public withLastName(name: string): UserBuilder {
        this._lastName = name;
        return this;
    }

    public withAge(age: number): UserBuilder {
        this._age = age;
        return this;
    }

    public liveInStreet(street: string): UserBuilder {
        this._street = street;
        return this;
    }

    public build(): User {
        let user: User = new User(this._firstName,
                              this._lastName,
                              this._age,
                              this._street);
        
        return user;
    }


}